File: backend/src/test/java/dev/coldhands/pair/stairs/backend/usecase/UserDetailsServiceTest.java
package dev.coldhands.pair.stairs.backend.usecase;

import dev.coldhands.pair.stairs.backend.domain.User;
import dev.coldhands.pair.stairs.backend.domain.UserName;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.UserEntity;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.AutoConfigureTestEntityManager;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.boot.test.context.SpringBootTest;

import java.time.Duration;
import java.time.Instant;
import java.util.UUID;

import static java.time.temporal.ChronoUnit.SECONDS;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;

@SpringBootTest
@AutoConfigureTestDatabase
@AutoConfigureTestEntityManager
@Transactional
class UserDetailsServiceTest {

    @Autowired
    private UserDetailsService underTest;

    @Autowired
    private TestEntityManager testEntityManager;

    @Test
    void createUserDetailsWhenUserDoesNotExist() {
        var userName = new UserName(
                null,
                null,
                "Jamie Redding"
        );
        var oidcSub = UUID.randomUUID().toString();

        final User user = underTest.createOrUpdate(oidcSub, userName);

        final UserEntity userEntity = testEntityManager.find(UserEntity.class, user.id());

        assertThat(userEntity).isNotNull();
        assertThat(userEntity.getId()).isEqualTo(user.id());
        assertThat(userEntity.getOidcSub()).isEqualTo(oidcSub);
        assertThat(userEntity.getDisplayName()).isEqualTo("Jamie");
        assertThat(userEntity.getCreatedAt()).isCloseTo(Instant.now(), within(Duration.of(1, SECONDS)));
        assertThat(userEntity.getUpdatedAt()).isCloseTo(Instant.now(), within(Duration.of(1, SECONDS)));
    }

    @Test
    void updateUserDetailsWhenUserDoesExist() {
        var oidcSub = UUID.randomUUID().toString();

        final User user = underTest.createOrUpdate(oidcSub, new UserName(
                null,
                null,
                "Jamie Redding"
        ));
        final UserEntity initialEntity = testEntityManager.find(UserEntity.class, user.id());
        final Instant initialCreatedAt = initialEntity.getCreatedAt();
        final Instant initialUpdatedAt = initialEntity.getUpdatedAt();

        underTest.createOrUpdate(oidcSub, new UserName(
                "Jay",
                null,
                "Jamie Redding"
        ));

        final UserEntity userEntity = testEntityManager.find(UserEntity.class, user.id());

        assertThat(userEntity).isNotNull();
        assertThat(userEntity.getId()).isEqualTo(user.id());
        assertThat(userEntity.getOidcSub()).isEqualTo(oidcSub);
        assertThat(userEntity.getDisplayName()).isEqualTo("Jay");
        assertThat(userEntity.getCreatedAt()).isEqualTo(initialCreatedAt);
        assertThat(userEntity.getUpdatedAt()).isAfter(initialUpdatedAt);
    }
}

File: backend/src/test/kotlin/dev/coldhands/pair/stairs/backend/usecase/UserDetailsServiceTest.kt
package dev.coldhands.pair.stairs.backend.usecase

import dev.coldhands.pair.stairs.backend.anOidcSub
import dev.coldhands.pair.stairs.backend.domain.RealDateProvider
import dev.coldhands.pair.stairs.backend.domain.User
import dev.coldhands.pair.stairs.backend.domain.UserName
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.dao.FakeUserDao
import dev.forkhandles.result4k.kotest.shouldBeSuccess
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.time.temporal.ChronoUnit

class UserDetailsServiceTest {

    private val userDao = FakeUserDao(RealDateProvider(), ChronoUnit.MILLIS)
    private val underTest: UserDetailsService = UserDetailsService(userDao)

    @Nested
    inner class CreateOrUpdate {

        @Test
        fun `create user details when user does not exist`() {
            val userName = UserName(
                nickName = null,
                givenName = null,
                fullName = "Jamie Redding",
            )
            val inputOidcSub = anOidcSub()

            val actualUser: User = underTest.createOrUpdate(inputOidcSub, userName).shouldBeSuccess()

            userDao.findById(actualUser.id).shouldNotBeNull {
                id shouldBe actualUser.id
                oidcSub shouldBe inputOidcSub
                displayName shouldBe "Jamie"
            }
        }

        @Test
        fun `update user details when user does exist`() {
            val inputOidcSub = anOidcSub()

            val initialUser = underTest.createOrUpdate(
                inputOidcSub,
                UserName(nickName = null, givenName = null, fullName = "Jamie Redding")
            ).shouldBeSuccess()

            val updatedUser = underTest.createOrUpdate(
                inputOidcSub,
                UserName(
                    nickName = "Jay",
                    givenName = null,
                    fullName = "Jamie Redding",
                )
            ).shouldBeSuccess()

            initialUser.id shouldBe updatedUser.id

            userDao.findById(initialUser.id).shouldNotBeNull {
                id shouldBe initialUser.id
                oidcSub shouldBe inputOidcSub
                displayName shouldBe "Jay"
            }
        }
    }
}